specification/flag-evaluation/flag-evaluation.md

```diff
@@ -0,0 +1,195 @@
+# Flag Evaluation API
+
+## Overview
+
+The `evaluation API` allows for the retrieval of feature flag values, independent of any flag control plane or vendor. In the absence of a feature flag system implementation (a [provider](../provider/providers.md)) the `evaluation API` uses the "No-op provider", which simply returns the supplied default flag value.
+
+#### Initialization and Configuration
+
+##### Requirement 1.1
+
+> The `API`, and any state it maintains **SHOULD** exist as a global singleton, even in cases wherein multiple versions of the `API` are present at runtime.
+
+It's important that multiple instances of the `API` not be active, so that state stored therein, such as the registered `provider`, static global `evaluation context`, and globally configured `hooks` allow the `API` to behave predictably. This can be difficult in some runtimes or languages, but implementors should make their best effort that only a single instance of the `API` is used.
+
+##### Requirement 1.2
+
+> The `API` **MUST** provide a function to set the `provider`, which accepts an API-conformant `provider` implementation.
+
+```
+// example provider mutator
+OpenFeature.setProvider(new MyProvider());
+```
+
+see [provider](../provider//providers.md)
+
+##### Requirement 1.3
+
+> The `API` **MUST** provide a function for `hook` addition, which accepts an array of API-conformant `hooks`. This adds the specified hooks to any hooks already attached.
+
+```
+// example hook attachment
+OpenFeature.addHooks([new MyHook()]);
+```
+
+see [hooks](./hooks.md)
+
+##### Requirement 1.4
+
+> The API **MUST** provide a function for retrieving the `provider` implementation.
+
+```
+// example provider accessor
+OpenFeature.getProvider();
+```
+
+see [provider](../provider/provider.md)
+
+##### Requirement 1.5
+
+> The `API` **MUST** provide a function for creating a `client`, with an options object containing an optional field for the client `name`.
+
+```
+// example client creation and retrieval
+OpenFeature.getClient({
+  name: 'my-openfeature-client'
+});
+```
+
+The name is a logical identifier for the client.
+
+##### Requirement 1.6
+
+> The client **MUST** provide a method for `hook` addition, which accepts an array of API-conformant `hooks`. This adds the specified hooks to any hooks already attached.
+
+```
+// example hook attachment
+client.addHooks([new MyHook()]);
+```
+
+see [hooks](./hooks.md)
+
+#### Flag Evaluation
+
+##### Requirement 1.7
+
+> The `client` **MUST** provide methods for typed flag value evaluation, including boolean, numeric, string, and map (object), with parameters `flag key` (string, required), `default value` (boolean | number | string | object, required), `evaluation context` (optional), and `evaluation options` (optional).
```

Comment:
Is the `boolean | number | string | object` syntax clear?

Reply:
Not at all clear to the uninitiated what "typed flag value evaluation" is

Reply:
See my comments on addHooks about language agnosticism. Map/object may not apply to all languages you want to target, and many languages differentiate between Map and object.  I would name and define this concept, and let any language use whatever mechanism is appropriate. Is it a map from string keys to (number | string | boolean)? Can the object have nested objects? Can the object have non-string keys? Can the values be arrays? If so, must those arrays be homogeneous?

Reply:
I think it might be best to say something suggesting that the language use whatever construct is usually used to interrogate and manipulate JSON objects, that's really what these are in most backends.

@justinabrahms @dyladan what do you think?

Reply:
JSON or arbitrary structured data? Could easily have a system which transmits these in YAML for example. Would that provider be expected to convert to a JSON specific structure?

```suggestion
> The `client` **MUST** provide methods for typed flag value evaluation, including boolean, numeric, string, and structured data, with parameters `flag key` (string, required), `default value` (boolean | number | string | object, required), `evaluation context` (optional), and `evaluation options` (optional). Structured data **SHOULD** be presented however is idiomatic in the implementation language for arbitrarily structured data such as JSON or YAML.
```